docs/enhancements/infrastructure-monitoring-service-api/alarms.md

```diff
@@ -0,0 +1,797 @@
+
+## Init
+Notes: 
+- Interacting with the following: `versions`, `alarm_dictionary` and `alarm_definitions`
+- `alarm_dictionary` has a primary key `resource_type_id`, ensuring that each ResourceType has a unique AlarmDictionary.
+- There's a one-to-many relation between `alarm_dictionary` and `alarm_definitions` via major.minor OCP version. The referential integrity and normalization version is done through `versions` table
+
+```sql
+INSERT INTO versions (version_number)
+VALUES
+    ('4.16'),
+    ('4.17');
+
+-- Most of the data in this table will come client-go calls
+INSERT INTO alarm_dictionary (alarm_dictionary_version, resource_type_id, entity_type, vendor)
+VALUES
+    ('4.16', 'b3e7149e-d471-4d0f-aaa6-d5e9aa9e713a', 'telco-model-OpenShift-4.16.2', 'Red Hat'),
+    ('4.16', '481688c8-2782-4534-a9de-88ca5154411d', 'telco-model-OpenShift-4.16.8', 'Red Hat' ),
+    ('4.17', 'f8b9e100-fd9f-4923-b96f-89418e9c2560', 'telco-model-OpenShift-4.17.2', 'Red Hat');
+
+-- Insert into alarm_definitions
+INSERT INTO alarm_definitions (alarm_name, alarm_last_change, alarm_description, proposed_repair_actions , alarm_additional_fields)
+VALUES
+    ('NodeClockNotSynchronising','4.16', 'Clock not synchronising.','https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/NodeClockNotSynchronising.md','{"customKey": "customValue"}'),
+    ('NodeClockNotSynchronising','4.17','Clock not synchronising.','https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/NodeClockNotSynchronising.md', '{"customKey": "customValue"}');
+-- The data here mapped from PrometheusRule. See ##PrometheusRule to AlarmDefinition mapping section for more.
+
+-- probable_cause will be auto populated. see the section "##Get Probable cause ID, name and description" for more details
+```
+## Detailed Server Instantiation
+
+For a given OCP release, the alarmDefinitions and probableCauses are fixed, so these can be built up front. For CaaS alarms only one resource type, “NodeCluster”, all alarms map to it.
+
+1. Query all managed clusters to get list of unique major.minor versions
+   - Need to monitor for major.minor new versions
+   - Audit on restarts
+
+2. For each major.minor version query PromRules from one cluster of that version
+
+3. Take the rules and build the corresponding alarmDefinition dictionary
+   - Link alarm dictionary to the resourceType
+   - alarmDictionary version will be that major.minor of cluster
+
+4. Assuming DB table already created, persist each alarmDictionary as a DB table
+
+5. Take all the PromRules and build the AlarmDefinition. See [here](#prometheusrule-to-alarmdefinition-mapping) for mapping. 
+   - A corresponding probablyCause row will be autometically added.
+
+6. Apply the required CR to activate the internal endpoint for alertmanager notification. See [here](#steps-for-internalv1caas-alertsalertmanager) for more.
+
+7. The server should now be ready to take requests. 
+
```

Comment:
I know it is implicit above, but I think we should clearly say somewhere that we will assume rules of all resources of a specific resource type are the same. This means we won't be able to see added rules to a specific resource, unless this one is selected for inspection (not deterministic).

Also, inspired from https://github.com/openshift-kni/oran-o2ims/pull/262, we could/should implement and sync mechanism for the alarm dictionary. This way we could overcome the limitation of not seeing newly added (or deleted) rules. In the documentation, we could say that once the system is up, rule modifications should be applied to all resources of a specific resource type and that the system will pick up these changes every X (i.e. 5 minutes) time.